hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+        return combineFees(List.of(defaultAirdropFees, cryptoTransferFees, tokenAssociationFees));
+    }
+
+    private Fees calculateCryptoTransferFees(
+            @NonNull FeeContext feeContext, @NonNull List<TokenTransferList> tokenTransfers) {
+        var cryptoTransferBody = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(tokenTransfers)
+                .build();
+
+        final var syntheticCryptoTransferTxn = TransactionBody.newBuilder()
+                .cryptoTransfer(cryptoTransferBody)
+                .transactionID(feeContext.body().transactionID())
+                .build();
+
+        return feeContext.dispatchComputeFees(syntheticCryptoTransferTxn, feeContext.payer());
+    }
+
+    /**
+     * Calculate the fees for the token associations that need to be created as part of the airdrop. It gathers all the
+     * token associations that need to be created and calculates the fees for each token association.
+     */
+    private Fees calculateTokenAssociationFees(FeeContext feeContext, TokenAirdropTransactionBody op) {
+        // Gather all the token associations that need to be created
+        var tokenAssociationsMap = new HashMap<AccountID, Set<TokenID>>();
+        final var tokenRelStore = feeContext.readableStore(ReadableTokenRelationStore.class);
+        for (var transferList : op.tokenTransfers()) {
+            final var tokenToTransfer = transferList.token();
+            for (var transfer : transferList.transfers()) {
+                if (tokenRelStore.get(transfer.accountID(), tokenToTransfer) == null) {
+                    var list = tokenAssociationsMap.getOrDefault(transfer.accountID(), new HashSet<>());
+                    list.add(tokenToTransfer);
+                    tokenAssociationsMap.put(transfer.accountID(), list);
+                }
+            }
+            for (var nftTransfer : transferList.nftTransfers()) {
+                if (tokenRelStore.get(nftTransfer.receiverAccountID(), tokenToTransfer) == null) {
+                    var list = tokenAssociationsMap.getOrDefault(nftTransfer.receiverAccountID(), new HashSet<>());
+                    list.add(tokenToTransfer);
+                    tokenAssociationsMap.put(nftTransfer.receiverAccountID(), list);
+                }
+            }
+        }
+
+        // Calculate the fees for each token association
+        var feeList = new ArrayList<Fees>();
+        for (var entry : tokenAssociationsMap.entrySet()) {
+            final var tokenAssociateBody = TokenAssociateTransactionBody.newBuilder()
+                    .account(entry.getKey())
+                    .tokens(new ArrayList<>(entry.getValue()))
+                    .build();
+
+            final var syntheticTxn = TransactionBody.newBuilder()
+                    .tokenAssociate(tokenAssociateBody)
+                    .transactionID(feeContext.body().transactionID())
```

Comment:
Is there a reason to set this here?

Reply:
Without it, throws an error that it's a null